### PR TITLE
[rebased] Added filters (sort and search) on manifest and build listing

### LINKED ISSUE #14: List all manifests with filters
- [x] List manifests
- [x] Sort by: 
  - [x] Name
  - [x] Creation date
  - [x] Last update date
- [x] Search by name

### LINKED ISSUE #24: List all builds with filters
- [x] List all builds
- [x] Sort by 
  - [x] Creation date
  - [x] Starting date
  - [x] Ending date
- [x] Search by
  - [x] Queuing
  - [x] Running
  - [x] Manifest id
  - [x] Exit status

### PR DESCRIPTION
**Pull request by BourgoisMickael**
Closes #14
Closes #24

For manifest listing:
sort by (name, creation update) and search by name

For build listing:
sort by (creation, start, end) and search by queuing, running, manifest_id and exit_status

### CODE REVIEW

File: server/src/action/build/list.js [hunk 1]
```diff
@@ -1,12 +1,52 @@
...
+      sanitizeQuery('exit_status')
+        .toInt(),
+      query('sort')
+        .optional()
+        .isString().withMessage('must be a string')
+        .isIn(['creation', 'start', 'end']).withMessage('sort by creation, start, end'),
+      query('dir')
+        .optional()
+        .isString().withMessage('must be a string')
+        .isIn(['asc', 'desc']).withMessage('direction is asc or desc')
```
Comment (Arignir) [inline server/src/action/build/list.js]:
```suggestion
        .isIn(['asc', 'desc']).withMessage('direction should either be "asc" or "desc"')
```

File: server/src/action/build/list.js [hunk 2]
```diff
@@ -1,12 +1,52 @@
...
+      sanitizeQuery('running')
+        .toBoolean(),
+      sanitizeQuery('manifest_id')
+        .toInt(),
+      sanitizeQuery('exit_status')
+        .toInt(),
+      query('sort')
+        .optional()
+        .isString().withMessage('must be a string')
+        .isIn(['creation', 'start', 'end']).withMessage('sort by creation, start, end'),
```
Comment (Arignir) [inline server/src/action/build/list.js]:
```suggestion
        .isIn(['creation', 'start', 'end']).withMessage('sort should either be "creation", "start" or "end"'),
```

File: server/src/action/build/list.js [hunk 3]
```diff
@@ -1,12 +1,52 @@
...
   async handler (req, res, next) {
-    return this.app.controller.build.list()
+    const sortToColumnName = {
+      'creation': 'creation_date',
+      'start': 'start_date',
+      'end': 'end_date'
+    }
+    // Default filters
+    const sort = sortToColumnName[req.query.sort || 'creation']
+    // Default direction is DESCENDING only if sort by creation
```
Comment (Arignir) [inline server/src/action/build/list.js]:
```suggestion
    // Default direction is DESCENDING only if we are sorting by creation date
```

File: server/src/action/build/list.js [hunk 4]
```diff
@@ -1,12 +1,52 @@
...
+    const dir = req.query.dir || (sort === 'creation_date' ? 'desc' : 'asc')
+    const filters = {
+      queuing: req.query.queuing || null,
```
Comment (Arignir) [inline server/src/action/build/list.js:43]: Why `|| null`? If they are absent, don't they already have for value `null`?

File: server/src/action/manifest/list.js [hunk 5]
```diff
@@ -1,12 +1,39 @@
...
-    return this.app.controller.manifest.list()
+    const sortToColumnName = {
+      'name': 'name',
+      'creation': 'creation_date',
+      'update': 'last_update'
+    }
+    // Default filters
+    const name = req.query.name || ''
+    const sort = sortToColumnName[req.query.sort || 'creation']
+    // Default direction is DESCENDING only if sort by creation
```
Comment (Arignir) [inline server/src/action/manifest/list.js]:
```suggestion
    // Default direction is DESCENDING only if we are sorting by creation date
```

File: server/src/controller/manifest.js [hunk 6]
```diff
@@ -71,9 +71,10 @@ class ManifestController {
     return manifest.toJSON()
   }
 
-  // TODO: filters
-  async list () {
+  async list (name, sort, direction) {
     const manifests = await this.app.database.model.manifest
+      .where('name', 'LIKE', name + '%')
```
Comment (Arignir) [inline server/src/controller/manifest.js:76]: Unsure as I'm not experienced with any techno used here, but the extra `%` makes me doubt: is this inserted as-is in the SQL query? In other words, is this field SQL-injectable using `name`?

**Review by Arignir — CHANGES_REQUESTED** [review]

File: server/src/action/build/list.js (hunk 4 above)
Comment (BourgoisMickael) [reply server/src/action/build/list.js:43]: No, they are `undefined`

File: server/src/controller/manifest.js (hunk 6 above)
Comment (BourgoisMickael) [reply server/src/controller/manifest.js:76]: `Bookshelf` uses 'Knex' a query builder which uses query bindings to protect against sql injection by sending parameters and query string separately to database driver, as explained here:[https://stackoverflow.com/questions/49665023/does-knex-js-prevent-sql-injection](https://stackoverflow.com/questions/49665023/does-knex-js-prevent-sql-injection)

So the only way to enable sql injection is by writing
```javascript
knex.raw(`select * from foo where id = ${id}`) // NEVER DO THIS 
```
But the methods used here always uses bindings format, to see that we can set `debug: true` in the knex configuration under `server/src/database/bookshelf.js` and the query will be logged:

```javascript
{ method: 'select',
  options: {},
  timeout: false,
  cancelOnTimeout: false,
  bindings: [ 'thisIsATest%' ],
  __knexQueryUid: '6ab1e80b-7af1-457d-8b53-362f24b580df',
  sql: 'select "manifest".* from "manifest" where "name" like ? order by "manifest"."creation_date" desc' }

```

File: server/src/action/build/list.js (hunk 4 above)
Comment (Arignir) [reply server/src/action/build/list.js:43]: That makes sense

**Review by Arignir — APPROVED** [review]

**Pull request merged**